File: Blackjack/Game.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;

namespace Blackjack2
{
    internal class Game
    {
        public int min_bet = 5;

        public int Bet(Player player)
        {
            int bet = 0;
            do
            {
                Console.WriteLine($"You currently have ${player.money}");
                Console.Write("How much would you like to bet? $");

                if (!int.TryParse(Console.ReadLine(), out bet))
                    Console.WriteLine("Invalid amount entered\n");
                else if (bet > player.money)
                    Console.WriteLine("You do not have that much money!\n");
                else if (bet <= 5)
                    Console.WriteLine($"You must enter a valid bet (minimum ${min_bet})!\n");
                else
                    Console.WriteLine();
            } while (bet > player.money || bet < min_bet);      //reprompts user if an invalid input is entered

            Console.Clear();
            return bet;
        }

        public void Deal(Player player, Deck d1, bool hide_card)     //deals cards to the player's hand
        {
            Console.WriteLine($"{player.name} has: ");
            player.hand[player.num_of_cards] = d1.SendCard();   //gets first card
            player.hand[player.num_of_cards].Draw_Card();       //displays first card
            player.num_of_cards++;                  //increments the number of cards in the player's hand

            player.hand[player.num_of_cards] = d1.SendCard();   //gets the second card
            if(hide_card)
                player.hand[player.num_of_cards].Draw_Back();
            else
                player.hand[player.num_of_cards].Draw_Card();       //displays the second card
            
            player.num_of_cards++;                  //increments the number of cards in the player's hand

            Thread.Sleep(3000);
            Console.Clear();                      //clears screen
        }

        public bool Choice(Player player, Deck d1, int bet)
        {
            bool cont = true;       //continues
            bool split = false;     //becomes true when the split function is chosen
            while (cont == true)
            {
                player.total = 0;   //resets player's total to 0 before it is recalculated
                int choice = 0;     //player's choice (entered later)
                bool flag = false;  //true if cards in hand total over 21
                cont = false;       //resets the continue flag
                int x = 0;          //counting variable
                Console.Clear();      //clears screen
                Console.WriteLine($"{player.name}, you have the: ");
                while (x < player.num_of_cards)     //displays player's cards and calculates the player's total
                {
                    Console.Write("\t");
                    player.hand[x].Display_Card();
                    Console.WriteLine();
                    player.total += player.hand[x].Get_Value(flag);
                    x++;
                }
                if (player.total > 21)  //recalculates total if ace must be set to 1 instead of 11
                {
                    x = 0;                  //resets counting variable
                    player.total = 0;       //resets total for new count 
                    while (x < player.num_of_cards)
                    {
                        flag = true;
                        player.total += player.hand[x].Get_Value(flag);
                        x++;
                    }
                }
                if (player.total > 21)  //recalculates total if ace must be set to 1 instead of 11
                {
                    x = 0;                  //resets counting variable
                    player.total = 0;       //resets total for new count 
                    while (x < player.num_of_cards)
                    {
                        flag = true;
                        player.total += player.hand[x].Get_Value(flag);
                        x++;
                    }
                }

                Console.WriteLine();
                Console.WriteLine($"You have a total of: {player.total}");
                Console.WriteLine("\n");

                if (player.total <= 21 && player.num_of_cards < 5)  //exits when player is over 21, or over 5 cards
                {
                    if (player.hand[0].rank == player.hand[1].rank && player.num_of_cards == 2 && player.money >= (bet * 2))  //allows split and double down options
                    {
                        if (player.total == 10 || player.total == 11)
                        {
                            do
                            {
                                Console.WriteLine($"Would you like to (1)Hit, (2)Stand, (3)Split, or (4)Double Down? ");    //displays options
                                choice = GetInput();
                                Console.WriteLine();
                                if (choice < 1 || choice > 4)
                                    Console.WriteLine($"Please make another selection\n");
                            } while (choice < 1 || choice > 4);     //continues on invalid selections
                        }
                        else    //allows split option
                        {
                            do
                            {
                                Console.WriteLine($"Would you like to (1)Hit, (2)Stand, or (3)Split? ");            //displays options
                                choice = GetInput();
                                Console.WriteLine();
                                if (choice < 1 || choice > 3)
                                    Console.WriteLine($"Please make another selection\n");
                            } while (choice < 1 || choice > 3);     //continues on invalid inputs
                        }
                    }
                    else if (player.hand[0].rank != player.hand[1].rank && (player.total == 10 || player.total == 11) && player.num_of_cards == 2)  //allows double down option
                    {
                        do
                        {
                            Console.WriteLine($"Would you like to (1)Hit, (2)Stand, or (3)Double Down? ");      //displays options
                            choice = GetInput();
                            Console.WriteLine();
                            if (choice < 1 || choice > 3)
                                Console.WriteLine($"Please make another selection\n");
                        } while (choice < 1 || choice > 3); //continues on invalid selection
                        if (choice == 3)
                            choice = 4;     //sets proper value to double down selection, to prevent from calling split function
                    }
                    else
                    {
                        do
                        {
                            Console.WriteLine($"Would you like to (1)Hit or (2)Stand? ");       //normal options
                            choice = GetInput();
                            Console.WriteLine();
                            if (choice < 1 || choice > 2)
                                Console.WriteLine($"Please make another selection\n");
                        } while (choice < 1 || choice > 2); //continues on invalid input
                    }
                }
                else
                    choice = 2;

                if (choice == 1)
                {
                    Hit(player, d1);        //calls hit function if selected
                    cont = true;
                }
                else if (choice == 3)
                {
                    split = true;           //calls split function if chosen
                    cont = false;
                }
                else if (choice == 4)
                {
                    DoubleDown(player, d1, bet);    //calls double down function if chosen
                    cont = false;
                }
            }
            return split;       //returns true if split is selected
        }

        private int GetInput()
        {
            int validInput = 0;
            var input = Console.ReadLine();
            if (int.TryParse(input, out validInput))
                return validInput;
            else
                return 0;
        }

        public void Hit(Player player, Deck d1)
        {
            Console.Clear();      //clears screen

            Console.WriteLine($"{player.name} received the ");
            player.hand[player.num_of_cards] = d1.SendCard();   //gets a new card from the deck
            player.hand[player.num_of_cards].Draw_Card();       //displays new card
            player.num_of_cards++;          //increments number of cards in hand

            Thread.Sleep(2000);   //waits 2 seconds
            Console.Clear();
        }

        public void Split(Player player, Player house, Deck d1, int bet)
        {
            Player player2 = new Player(player.name + "'s second hand", 0);       //new object, player's second hand
            player.num_of_cards = 1;
            player2.num_of_cards = 1;
            player2.hand[0] = player.hand[1];   //splits hand
            Hit(player, d1);                    //adds a new card to first hand
            Hit(player2, d1);                   //adds a new card to second hand

            Choice(player, d1, bet);            //plays first hand
            Choice(player2, d1, bet);           //plays second hand

            H_Play(house, d1);                  //house plays

            Console.WriteLine($"First Hand:");
            Decision(player, house, d1, bet);   //decides if first hand wins
            Console.WriteLine($"Second Hand:");
            Decision(player2, house, d1, bet);  //decides if second hand wins
            player.money += (player2.money);

            Settle(player, house, bet);     //resets values from next hand
        }

        public void DoubleDown(Player player, Deck d1, int bet)
        {
            bool x = false;
            Hit(player, d1);    //gives the player one more card
            bet *= 2;           //doubles the bet
            if (player.total == 11)
                x = true;
            player.total += player.hand[2].Get_Value(x);
        }

        public void H_Play(Player house, Deck d1)        //house plays its hand
        {
            bool cont = true;       //continues until house is finished
            while (cont == true)
            {
                house.total = 0;
                bool flag = false;
                cont = false;
                int x = 0;          //counter variable
                Console.Clear();      //clear screen
                Console.WriteLine($"{house.name} has the: ");
                while (x < house.num_of_cards)      //displays house's cards and calculates the total value
                {
                    Console.WriteLine("\t");
                    house.hand[x].Display_Card();       //displays the cards that are in the house's hand
                    Console.WriteLine();
                    house.total += house.hand[x].Get_Value(flag);       //calculates the total value
                    x++;            //increment counter
                }
                if (house.total > 21)       //recounts with ace equal to 1
                {
                    x = 0;
                    house.total = 0;
                    while (x < house.num_of_cards)
                    {
                        flag = true;
                        house.total += house.hand[x].Get_Value(flag);
                        x++;
                    }
                }

                Console.WriteLine();
                Console.Write($"{house.name} has a total of: {house.total}, and chooses to ");
                if (house.total < 17)
                {
                    Console.Write("hit\n");
                    Thread.Sleep(2500);   //wait 2.5 seconds
                    Hit(house, d1);                 //deals another card to the house
                    cont = true;
                }
                else
                    Console.Write("stand\n");
            }
        }

        public void Decision(Player player, Player house, Deck d1, int bet)
        {
            Console.WriteLine($"You have a total of {player.total}\n");

            Thread.Sleep(2000);       //wait 2 seconds

            if (player.total < 21 && player.num_of_cards == 5 && house.num_of_cards != 5)   //if you win five card rule
            {
                Console.WriteLine("You Win! (five card rule)\n");
                player.money += bet;
            }
            else if (house.total < 21 && player.num_of_cards == 5 && player.num_of_cards != 5)  //if you lose five card rule
            {
                Console.WriteLine("You Lose (five card rule)\n");
                player.money -= bet;
            }
            else
            {
                if (player.total == 21 && player.num_of_cards == 2) //if you have a blackjack
                {
                    if (house.total != 21 || house.num_of_cards > 2)    //if you win
                    {
                        Console.WriteLine("You have a BlckJack!\nYou Win!\n");
                        player.money += (bet * 1.5);
                    }
                    else                                                //if house has one too
                    {
                        Console.WriteLine($"Draw\n");
                    }
                }
                else if (house.total == 21 && house.num_of_cards == 2)  //if house has a blackjack
                {
                    Console.WriteLine("The House has a BlckJack!\nYou Lose\n");
                    player.money -= bet;
                }
                else
                {
                    if (player.total > 21)                      //test player for bust
                        Console.WriteLine("You are over!");
                    if (house.total > 21)                       //test house for bust
                        Console.WriteLine("House is over!");
                    if (player.total > 21 && house.total > 21)  //test both for bust
                    {
                        Console.WriteLine("Draw\n");
                    }
                    else if (player.total > 21 && house.total <= 21)    //if player is over
                    {
                        Console.WriteLine("You Lose\n");
                        player.money -= bet;
                    }
                    else if (house.total > 21 && player.total <= 21)    //if house is over
                    {
                        Console.WriteLine("You Win!\n");
                        player.money += bet;
                    }
                    else if (house.total <= 21 && player.total <= 21)   //both are under
                    {
                        if (house.total < player.total)                 //if player beats house in points
                        {
                            Console.WriteLine("You Win!\n");
                            player.money += bet;
                        }
                        else if (house.total > player.total)            //if house beats player in points
                        {
                            Console.WriteLine("You Lose\n");
                            player.money -= bet;
                        }
                        else                                            //if tie with points
                        {
                            if (player.num_of_cards < house.num_of_cards)   //if player has fewer cards than house
                            {
                                Console.WriteLine($"You Win! (Fewer Cards)\n");
                                player.money += bet;
                            }
                            else if (player.num_of_cards > house.num_of_cards)  //if player has more cards than house
                            {
                                Console.WriteLine($"You Lose (House has fewer cards)\n");
                                player.money -= bet;
                            }
                            else                                                //both are equal (draw)
                                Console.WriteLine($"Draw\n");
                        }
                    }
                }
            }
        }

        public void Settle(Player player, Player house, int bet)
        {
            player.num_of_cards = 0;        //resets player's hand
            house.num_of_cards = 0;         //resets house's hand
            bet = 0;                        //resets bet
        }

        public char Go_Play()
        {
            char go;        //variable to continue
            Console.WriteLine("Would you like to continue(y/n)? ");
            go = Console.ReadKey().KeyChar;
            Console.Clear();      //clear screen
            return go;          //returns user's input
        }
    }
}

File: PythonBlackjack/Game.py
import random
import time
import math

from Player import Player
from Deck import Deck
from GameDisplay import GameDisplay

class Game:
    def __init__(self, player=None, house=None, deck=None, game_display=None):
        self.__min_bet = 5
        self.__current_bet = 0
        self._player = player if player is not None else Player()
        self._house = house if house is not None else Player("The House", starting_amount=0)
        self._deck = deck if deck is not None else Deck()
        if(game_display != None and isinstance(game_display, GameDisplay)):
            self._game_display = game_display  
        else:
            self._game_display = GameDisplay()
        self._deck.shuffle()

    @property
    def min_bet(self):
        return self.__min_bet
    
    def play(self):
        while self._player.money >= self.__min_bet:
            if self._deck.position > 40:
                self._deck.shuffle()

            self.__current_bet = self.bet()
            self.deal()
            split = self.play_hand(self._player)
            if split:
                self.split()
            else:
                self.h_play()
                result = self.decision(self._player)
                self.settle(result)  # No need for bet argument in settle

            if(self._player.money < self.__min_bet):
                break
            go = input("Continue (or x to exit)... ")
            print("\033c", end="")  # Clear screen
            if go.lower() == 'x':
                break
        
        print(f"Your final total was: ${self._player.money}")
        print("Thanks for playing!\n")

    def _validate_bet(self, bet):
        try:
            bet = int(bet)
        
            if bet > self._player.money:
                print("You do not have that much money!\n")
            elif bet < self.__min_bet:
                print(f"You must enter a valid bet (minimum ${self.__min_bet})!\n")
            else:
                print()  # Print a newline on successful input
                return True
        except ValueError:
            print("Invalid amount entered\n")
        return False
   
    def bet(self):
        while True:
            print(f"You have ${self._player.money}")
            bet = input("Enter bet: $")
            if self._validate_bet(bet):
                break
        return int(bet)

    def deal(self):
        self.hit(self._player, hole_card=True)
        self.hit(self._player, hole_card=True)
        
        self.hit(self._house, hole_card=True)
        self.hit(self._house, hole_card=True)
        
        self._game_display.print_game(self._player, self._house, self.__current_bet, True)

    def play_hand(self, player):
        cont = True
        split = False
        while cont:
            choice = 2
            cont = False
            
            option_split = False
            option_double = False

            if player.total <= 21 and player.num_of_cards < 5:

                if player.num_of_cards == 2 and player.money >= (self.__current_bet * 2):
                    if player.hand[0].rank == player.hand[1].rank:
                        option_split = True
                    elif player.total == 10 or player.total == 11:
                        option_double = True
                
                while True:
                    print("Would you like to (1)Hit, (2)Stand", end="")
                    if option_split and option_double:
                        print(", (3)Split, or (4)Double Down? ", end="")
                    elif option_split:
                        print(", or (3)Split? ", end="")
                    elif option_double:
                        print(", or (3)Double Down? ", end="")
                    else:
                        print("? ", end="")

                    max_options = 2
                    if option_split:
                        max_options += 1
                    if option_double:
                        max_options += 1
                    
                    choice = self.get_input()
                    print()
                    if 1 <= choice <= 4:
                        break
                    else:
                        print("Please make another selection\n")

            if choice == 1:
                self.hit(player, hole_card=True)
                self._game_display.print_game(player, self._house, self.__current_bet, True)
                cont = True
            elif choice == 3:
                if option_double and not option_split:
                    self.double_down(player)
                else:
                    split = True
                cont = False
            elif choice == 4:
                self.double_down(self.__current_bet)
                cont = False
        return split

    def get_input(self):
        try:
            return int(input())
        except ValueError:
            return 0

    def hit(self, player, hole_card=False):
        player.get_card(self._deck.get_top_card())
        if(player == self._player or player == self._house):
            self._game_display.print_game(self._player, self._house, self.__current_bet, hole_card)
        else:
            self._game_display.print_game(player, self._house, self.__current_bet, hole_card)

        time.sleep(.5)

    def split(self):
        player2 = Player(self._player.name + "(second)", 0)
        player2.hand.append(self._player.hand.pop())
        player1 = Player(self._player.name + "(first)", 0)
        player1.hand.append(self._player.hand.pop())
        
        self.hit(player1, hole_card=True)
        self.play_hand(player1)
        self.hit(player2, hole_card=True)
        self.play_hand(player2)
        self.h_play(player2)

        print(f"First Hand: {player1.total} - ", end="")
        result1 = self.decision(player1)
        print(f"Second Hand: {player2.total} - ", end="")
        result2 = self.decision(player2)
        self.settle(result1 + result2)

    def double_down(self, player):
        self.hit(player, hole_card=True)
        self.__current_bet *= 2

    def h_play(self, player=None):
        if player is None:
            player = self._player
        while True:
            self._game_display.print_game(player, self._house, self.__current_bet)
            time.sleep(.5)
            print(f"{self._house.name} chooses to ", end="")
            if self._house.total < 17:
                print("hit\n")
                time.sleep(.5)
                self._house.get_card(self._deck.get_top_card())
                cont = True
            else:
                print("stand\n")
                break

    def decision(self, player):
        time.sleep(.5)

        if player.total <= 21 and player.num_of_cards == 5 and self._house.num_of_cards!= 5:
            print("You Win! (five card rule)")
            return self.__current_bet
        if self._house.total <= 21 and self._house.num_of_cards == 5 and player.num_of_cards!= 5:
            print("You Lose (five card rule)")
            return (self.__current_bet * -1)
        if player.total == 21 and player.num_of_cards == 2:
            if self._house.total == 21 and self._house.num_of_cards == 2:
                print("Draw")
                return 0
            else:
                print("You have a Blackjack!  You Win!")
                return math.ceil(self.__current_bet * 1.5)
        if self._house.total == 21 and self._house.num_of_cards == 2:
            print("The House has a Blackjack! You Lose")
            return (self.__current_bet * -1)
        
        #player or house is over 21
        if player.total > 21:
            print("You are over! ", end="")
        if self._house.total > 21:
            print("House is over! ", end="")
        if player.total > 21 and self._house.total > 21:
            print("Draw")
            return 0
        if player.total > 21 and self._house.total <= 21:
            print("You Lose")
            return (self.__current_bet * -1)
        if self._house.total > 21 and player.total <= 21:
            print("You Win!")
            return self.__current_bet
        
        if self._house.total < player.total:
            print("You Win!")
            return self.__current_bet
        if self._house.total > player.total:
            print("You Lose")
            return (self.__current_bet * -1)
        # if self._house.total == player.total and player.num_of_cards < self._house.num_of_cards:
        #     print("You Win! (Fewer Cards)")
        #     return self.__current_bet
        # if self._house.total == player.total and player.num_of_cards > self._house.num_of_cards:
        #     print("You Lose (House has fewer cards)")
        #     return (self.__current_bet * -1)
        if self._house.total == player.total:
            print("Draw")
            return 0
        else:
            print("Error in decision logic")
            return 0

    def settle(self, result):
        print(f"Result: {result}")
        self._player.settle_bet(result)
        self._player.reset_hand()
        self._house.reset_hand()
        self.__current_bet = 0
        
